shared/health/src/main/java/io/pravega/shared/health/impl/ContributorRegistryImpl.java

```diff
@@ -0,0 +1,312 @@
+     * @return The {@link HealthContributor} registered under the {@link ContributorRegistry}.
+     */
+    @NonNull
+    @Override
+    public HealthContributor register(HealthComponent component, HealthComponent parent) {
+        // Use the default (root) component if parent is undefined.
+        parent = parent == null ? root : parent;
+        // A HealthComponent should only exist if defined during construction, instead of adding it dynamically.
+        if (!components.contains(parent.getName()) || !contributors.containsKey(parent.getName())) {
+            log.warn("Attempting to register {} under unrecognized {} -- aborting.", component, parent);
+            return null;
+        }
+        components.add(component.getName());
+        return register(component, parent.getName());
+    }
+
+    /**
+     * Registers a {@link HealthContributor} under the specified parent {@link HealthComponent}.
+     *
+     * @param contributor The {@link HealthContributor} object to add to the registry as a child of 'parent'.
+     * @param parent The {@link HealthComponent} to add 'contributor' as a child/dependency.
+     * @return The {@link HealthContributor} registered under the {@link ContributorRegistry}.
+     */
+    @NonNull
+    public HealthContributor register(HealthContributor contributor, HealthComponent parent) {
+        return register(contributor, parent.getName());
+    }
+
+    /**
+     * Registers a {@link HealthContributor} under the specified parent {@link HealthComponent}.
+     *
+     * @param contributor The {@link HealthContributor} object to add to the registry as a child of 'parent'.
+     * @param component The {@link String} id of a {@link HealthComponent} to add 'contributor' as a child/dependency.
+     * @return The {@link HealthContributor} registered under the {@link ContributorRegistry}. Null return values are permitted.
+     */
+    @NonNull
+    @Override
+    public HealthContributor register(HealthContributor contributor, String component) {
+        String name = contributor.getName();
+        // HealthContributor mapped by 'parent' should exist at time of some child registration.
+        if (!contributors.containsKey(component)) {
+            log.debug("Unrecognized HealthContributor::{} -- aborting registration.", component);
+            return null;
+        }
+        if (contributors.containsKey(contributor.getName())) {
+            log.warn("{} has already been registered -- aborting.", contributors.get(contributor.getName()));
+            return null;
+        }
+        log.debug("Registering {} to {}.", contributor, contributors.get(component));
+        contributors.putIfAbsent(name, contributor);
+        // A new contributor should have no child relations to overwrite.
+        // Realistically, this should be restricted because it implies non HealthComponent contributors can have children.
+        children.put(name, new HashSet<>());
```

Comment:
for these next few lines, just assign this HashMap to a local variable to prevent you from putting it and then immediately getting it out of the map.